### DISCUSSION TITLE
Custom allocator for a uniform buffer of SVector/SMatrix

### DISCUSSION THREAD

**Discussion by JMLX42** [dimforge/nalgebra#1201] [Q&A]
Hello @sebcrozet !

Incredible work on this project. I hope you're doing well 😄 

I am working on a new version of [the open source SmartShape Engine](https://github.com/aerys/minko) written 100% in Rust using [wgpu](https://github.com/gfx-rs/wgpu).

WebGPU has uniform buffers: the goal is to avoid many calls to `glUniform()` with a single call to set a uniform buffer. I would like to take it one step further.

I would like my Rust code to do math on Vector and Matrices that are allocated in a uniform buffer. Presumably a `Vec<u8>` or `[u8]`. The goal is to make sure the underlying memory for a specific set of Vector/Matrices is a contiguous buffer than can directly be used as a uniform buffer without copy.

I have read the [Computer-graphics recipes](https://nalgebra.org/docs/user_guide/cg_recipes) already. But the difference is it relies on copying the Vector/Matrix data to the uniform buffer.

My first take would be that I would need to implement a custom allocator for static `SVector` and `SMatrix`. Am I going in the right direction? Does it even make sense?

Thanks!

**Comment by sebcrozet** [comment]
Hey!

I think using a custom implementation of the `Allocator` traits would be too complex (if at all possible) since there is no way to override the `DefaultAllocator` used for the output of most operations in `nalgebra`. So there is no easy way to force any operation of `nalgebra` to output the result of any operation (that doesn’t operate in-place) into your own buffer. The `Allocator` system in `nalgebra` is mainly a way for us to deal with dimension-dependent storage, rather that something users can customize.

If all the elements of your uniform buffer are supposed to have the same type, you can enable the `bytemuck` feature of `nalgebra`, and use the `bytemuck` crate to convert your `&mut [u8]` to, for example, a `&mut [Vector3<f32>]`, and store the result of your math ops on that slice (or apply in-place math operations on these vectors).

**Reply by JMLX42** [reply]
@sebcrozet thank you for your answer!

> If all the elements of your uniform buffer are supposed to have the same type

IMHO that's unlikely. The primary driver to place a specific uniform into a specific buffer is most likely semantics. For example, group all the material properties in a single buffer, so I can set that specific uniform buffer once and do all the corresponding draw calls.

> store the result of your math ops on that slice (or apply in-place math operations on these vectors).

@sebcrozet what do you mean exactly by "apply in-place math operations"? Do you have an example with `nalgebra`?
